### PR TITLE
Test sketch for utilising serial callback library

### PR DESCRIPTION
**Pull request by marcelvanworkum**
This will allow us to get pretty much all the functionality we want right off the bat,
and apparently it integrates nicely with GUIs as well.

### CODE REVIEW

File: hardware/src/base-station-sketch2/CmdMessenger.cpp [hunk 1]
```diff
@@ -0,0 +1,680 @@
+/*
+  CmdMessenger - library that provides command based messaging
```
Comment (jamiesanson) [inline hardware/src/base-station-sketch2/CmdMessenger.cpp]: Did we decide to not include library source as MIT?

File: hardware/src/base-station-sketch2/sketch_aug16a.ino [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    current_status,
+    status_is,
+    think_for_me,
+    thought_is,
+    error,
+};
+
+/* Initialize CmdMessenger -- this should match PyCmdMessenger instance */
+const int BAUD_RATE = 9600;
+CmdMessenger c = CmdMessenger(Serial,',',';','/');
```
Comment (jamiesanson) [inline hardware/src/base-station-sketch2/sketch_aug16a.ino]: This set of arguments is extremely unintuitive

File: hardware/src/base-station-sketch2/sketch_aug16a.ino [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+/* Attach callbacks for CmdMessenger commands */
+void attach_callbacks(void) { 
+    c.attach(current_status,on_current_status);
+    c.attach(think_for_me,on_think_for_me);
+    c.attach(on_unknown_command);
+}
+
+void setup() {
+    Serial.begin(BAUD_RATE);
+    attach_callbacks();
```
Comment (jamiesanson) [inline hardware/src/base-station-sketch2/sketch_aug16a.ino]: This actually seems like a very elegant way of doing things, I like!

File: hardware/src/base-station-sketch2/sketch_aug16a.ino (hunk 2 above)
Comment (marcelvanworkum) [reply hardware/src/base-station-sketch2/sketch_aug16a.ino]: I agree, I am not even sure what they're doing to be fair. That was just copied from an example

File: hardware/src/base-station-sketch2/CmdMessenger.h [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+  */
+
+#ifndef CmdMessenger_h
+#define CmdMessenger_h
+
+#include <inttypes.h>
+#if ARDUINO >= 100
```
Comment (jamiesanson) [inline hardware/src/base-station-sketch2/CmdMessenger.h]: This library will need testing with teensy, as I'm not sure which version of arduino its equivalent to

File: hardware/src/base-station-sketch2/CmdMessenger.cpp (hunk 1 above)
Comment (marcelvanworkum) [reply hardware/src/base-station-sketch2/CmdMessenger.cpp]: This and the `.h` file need to be included in the same directory as the `.ino`. Not sure if there is a work around for that.

File: hardware/src/base-station-sketch2/CmdMessenger.h (hunk 4 above)
Comment (marcelvanworkum) [reply hardware/src/base-station-sketch2/CmdMessenger.h]: Do we have one available atm?

File: hardware/src/base-station-sketch2/CmdMessenger.cpp (hunk 1 above)
Comment (jamiesanson) [reply hardware/src/base-station-sketch2/CmdMessenger.cpp]: These files make up an Arduino "library" , so should follow the same conventions as we use of other libraries

File: hardware/src/base-station-sketch2/CmdMessenger.h (hunk 4 above)
Comment (jamiesanson) [reply hardware/src/base-station-sketch2/CmdMessenger.h]: I can grab one in Wednesday

File: hardware/src/base-station-sketch2/sketch_aug16a.ino (hunk 2 above)
Comment (jamiesanson) [reply hardware/src/base-station-sketch2/sketch_aug16a.ino]: Looked at the sauce for the constructor Def in `CmdMessenger.h` and they're actually unnecessary in this case. Remove all arguments other than `Serial` and it should work the same

**Comment by marcelvanworkum** [comment]
@jamiesanson I have refactored this pull request, could you take another look?

File: hardware/src/base-station-sketch/example-pycmdmessenger-sketch.ino [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  By Marcel van Workum
+*/
+
+#include "CmdMessenger.h"
+
+/* Define available CmdMessenger commands */
+enum {
+    current_status,
+    status_is,
+    think_for_me,
```
Comment (jamiesanson) [inline hardware/src/base-station-sketch/example-pycmdmessenger-sketch.ino]: These are kind of oddly named. I get that it's an example but it doesn't relate to rocket things

**Comment by marcelvanworkum** [comment]
@jamiesanson changed those commands to be more rocket oriented, anything else or good to merge?

**Comment by jamiesanson** [comment]
Looks good! Merging now

**Pull request merged**